module/zfs/spa.c

```diff
@@ -4294,6 +4306,8 @@ spa_export_common(char *pool, int new_state, nvlist_t **oldconfig,
 		zvol_remove_minors(spa, spa_name(spa), B_TRUE);
 		taskq_wait(spa->spa_zvol_taskq);
 	}
+	if (spa->spa_upgrade_taskq)
+		taskq_wait(spa->spa_upgrade_taskq);
```

Comment:
I don't think this is needed.  By the time we get to exporting all datasets should be disowned and therefore we've waited on any outstanding upgrades to exit.  The taskq should be empty.  If somehow that's not the case we could end up blocking for potentially a very long time.